src/scanner/mod.rs

```diff
@@ -130,8 +137,33 @@ impl Scanner {
     ///
     /// Note: `self` must contain `self.ip`.
     async fn scan_socket(&self, socket: SocketAddr) -> io::Result<SocketAddr> {
-        let tries = self.tries.get();
+        if self.sudp {
+            // we try different wait times and packet types
+            // I looked for a crate to create these packets but couldn't find one
+            let waits = vec![0, 50, 100, 300];
+            let payloads = vec![
+                craft_dns_query_packet(),
+                craft_ntp_packet(),
+                craft_dhcpc_packet(),
+            ];
 
+            for payload in &payloads {
+                for &wait_ms in &waits {
+                    let wait = Duration::from_millis(wait_ms);
+                    match self.udp_scan(socket, payload.clone(), wait).await {
```

Comment:
`payload.clone()`

I think you can improve here.

You reference payload &payloads on line 150, and then clone them in a double for loop.

I think this is a bit inefficient, you could instead clone here:

```
for payload in &payloads
```

I am thinking you do not even need to make payloads at all:

```

            let payloads = vec![
                craft_dns_query_packet(),
                craft_ntp_packet(),
                craft_dhcpc_packet(),
            ];
```

I am thinking you can just call the right function in `udp_scan` when it's needed. I have not read all of the code yet though 🙈

Reply:
Ahh ok reading your comment:

> Better news, looking more into nmap they are actually just sending empty packets to most ports on UDP scans, crafting specific payloads for specific well known ports.

Yes, I would look at this a bit more. Crafting 3 packets and hoping the server responds to one is a bit costly. By moving into empty packets you can speed up a lot